[question] CMakeConfigDeps: Mixing C and C++ fails

### What is your question?

Hi,

I tried to switch from `CMakeDeps` to `CMakeConfigDeps` for one of our libraries.
However, I run into some issue regarding enabled languages inside CMake.

When using some C library and trying to use that inside a C++ project, I get errors with `CMakeConfigDeps`, it works fine with `CMakeDeps`.
CMake itself has no problem with that mixed scenario, as it is very common.

#### How to trigger the error

The required library must have `languages = "C"` inside its `conanfile.py` (I used `capstone` in my example).
The consumer must use `project(<PROJECT_NAME> LANGUAGES CXX)` inside its `CMakeLists.txt`.

`conanfile.py`:
```py
from conan import ConanFile
from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout


class Consumer(ConanFile):
    settings = ("os", "compiler", "build_type", "arch")

    def requirements(self):
        # Any other library with languages = "C" would do as well,
        # e.g. cmake_lib from conan new when adding the languages line.
        self.requires("capstone/[>=5]")

    def layout(self):
        cmake_layout(self)

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()

        deps = CMakeDeps(self)
        deps.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
```

`CMakeLists.txt`:
```cmake
cmake_minimum_required(VERSION 3.23)
project(consumer LANGUAGES CXX)

find_package(capstone REQUIRED CONFIG)
```

#### Environment

Linux
Conan version: 2.26.1
No special profile.

#### Using `CMakeDeps`

```
$ conan build .
...
-- Conan: Target declared 'capstone::capstone'
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/consumer_cpp/build/x86_64-linux-gcc-release
```

#### Trying to use `CMakeConfigDeps`

```
$ conan build . -c tools.cmake.cmakedeps:new=will_break_next
...
-- Conan: Configuring Targets for capstone/5.0.7
-- Conan: Target declared imported SHARED library 'capstone::capstone'
CMake Error at build/x86_64-linux-gcc-release/generators/capstone-Targets-release.cmake:20 (message):
  Target capstone::capstone has C linkage but C not enabled in project()
Call Stack (most recent call first):
  build/x86_64-linux-gcc-release/generators/capstoneTargets.cmake:7 (include)
  build/x86_64-linux-gcc-release/generators/capstone-config.cmake:6 (include)
  CMakeLists.txt:4 (find_package)


-- Configuring incomplete, errors occurred!
```

#### Additional information

The relevant code is [this](https://github.com/conan-io/conan/blob/084f37d67ed2cbeb5f4857f1c46a7fa8e9db134f/conan/tools/cmake/cmakeconfigdeps/target_configuration.py#L412-L415), which was initially introduced via 5f75738.
I could not find any reasoning for that check.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

--- a/conan/tools/cmake/cmakeconfigdeps/target_configuration.py
+++ b/conan/tools/cmake/cmakeconfigdeps/target_configuration.py
@@ -408,6 +408,12 @@ class TargetConfigurationTemplate2:
 
         {% if lib_info.get("link_languages") %}
         get_property(_languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+        if("CXX" IN_LIST _languages)
+            list(APPEND _languages "C")
+        endif()
+        if("CUDA" IN_LIST _languages)
+            list(APPEND _languages "C" "CXX")
+        endif()
         {% for lang in lib_info["link_languages"] %}
         if(NOT "{{lang}}" IN_LIST _languages)
             message(SEND_ERROR

--- a/test/functional/toolchains/cmake/cmakedeps2/test_cmakeconfigdeps_new_cpp_linkage.py
+++ b/test/functional/toolchains/cmake/cmakedeps2/test_cmakeconfigdeps_new_cpp_linkage.py
@@ -2,8 +2,117 @@ import platform
 import textwrap
 import pytest
 
+from conan.test.assets.sources import gen_function_c, gen_function_h
+from conan.test.utils.tools import TestClient
 
-new_value = "will_break_next"
+
+@pytest.mark.tool("cmake")
+def test_cxx_only_project_links_c_library():
+    """
+    When the consumer enables only CXX (project(myapp CXX)), the generated CMake config
+    must not SEND_ERROR for a dependency with C linkage. CXX implies C linkage support,
+    so the template adds "C" to the enabled languages check when "CXX" is present.
+    """
+    c = TestClient()
+    # C library package: builds a static C lib, declares languages = "C" so link_languages = ["C"]
+    hello_c = gen_function_c(name="hello")
+    hello_h = gen_function_h(name="hello")
+    hello_cmake = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(hello C)
+        add_library(hello STATIC src/hello.c)
+        target_include_directories(hello PUBLIC
+          $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+          $<INSTALL_INTERFACE:include>)
+        install(TARGETS hello
+          ARCHIVE DESTINATION lib
+          INCLUDES DESTINATION include)
+        install(FILES include/hello.h DESTINATION include)
+        """)
+    hello_conanfile = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.cmake import CMake, cmake_layout
+
+        class Clib(ConanFile):
+            name = "hello"
+            version = "0.1"
+            settings = "os", "compiler", "build_type", "arch"
+            package_type = "static-library"
+            generators = "CMakeToolchain"
+            exports_sources = "CMakeLists.txt", "src/*", "include/*"
+            languages = "C"
+
+            def layout(self):
+                cmake_layout(self)
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure()
+                cmake.build()
+
+            def package(self):
+                cmake = CMake(self)
+                cmake.install()
+
+            def package_info(self):
+                self.cpp_info.libs = ["hello"]
+        """)
+    c.save({
+        "hello/conanfile.py": hello_conanfile,
+        "hello/CMakeLists.txt": hello_cmake,
+        "hello/src/hello.c": hello_c,
+        "hello/include/hello.h": hello_h,
+    })
+    c.run("create hello")
+
+    # Consumer: project(CXX) only - no C in project(). Links to C library.
+    consumer_cmake = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(myapp CXX)
+
+        find_package(hello CONFIG REQUIRED)
+
+        add_executable(app main.cpp)
+        target_link_libraries(app PRIVATE hello::hello)
+        """)
+    main_cpp = textwrap.dedent("""
+        extern "C" {
+            #include "hello.h"
+        }
+        int main() {
+            hello();
+        }
+        """)
+    consumer_conanfile = textwrap.dedent("""\
+        import os
+        from conan import ConanFile
+        from conan.tools.cmake import CMake, cmake_layout
+
+        class Recipe(ConanFile):
+            settings = "os", "compiler", "build_type", "arch"
+            package_type = "application"
+            generators = "CMakeToolchain", "CMakeConfigDeps"
+            requires = "hello/0.1"
+
+            def layout(self):
+                cmake_layout(self)
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure()
+                cmake.build()
+                self.run(os.path.join(self.cpp.build.bindir, "app"), env="conanrun")
+        """)
+    c.save({
+        "consumer/conanfile.py": consumer_conanfile,
+        "consumer/CMakeLists.txt": consumer_cmake,
+        "consumer/main.cpp": main_cpp,
+    })
+    c.run("build consumer")
+
+    # If the fix were missing, CMake configure would SEND_ERROR: "Target hello::hello has C
+    # linkage but C not enabled in project()". So reaching here and running the app proves the fix.
+    assert "hello: Release!" in c.out
 
 
 @pytest.mark.skipif(platform.system() == "Windows", reason="Windows doesn't fail to link")
@@ -28,7 +137,7 @@ def test_auto_cppstd(matrix_c_interface_client):
         class Recipe(ConanFile):
             settings = "os", "compiler", "build_type", "arch"
             package_type = "application"
-            generators = "CMakeToolchain", "CMakeDeps"
+            generators = "CMakeToolchain", "CMakeConfigDeps"
             requires = "matrix/0.1"
 
             def layout(self):
@@ -50,5 +159,5 @@ def test_auto_cppstd(matrix_c_interface_client):
     c.save({"conanfile.py": conanfile,
             "CMakeLists.txt": consumer,
             "app.c": app}, clean_first=True)
-    c.run(f"build . -c tools.cmake.cmakedeps:new={new_value}")
+    c.run(f"build .")
     assert "Hello Matrix!" in c.out

